feat: pnpm bundling

JeremyJonas opened the pull request:
- Adds transitive bundled dependencies symlinks to packages during package task
- Use `NX_NON_NATIVE_HASHER=true` for all run-many tasks
- Upgrade nx versions
- Use `--runInBand` for nx run-many jest executions

JeremyJonas committed:
chore(workflows): use `NX_NON_NATIVE-HASHER=true` for all workflows

Reverts default from https://github.com/nrwl/nx/pull/15071 which breaks
our caching.

JeremyJonas committed:
chore: upgrade nx

JeremyJonas committed:
chore(nx-monorepo): set `NX_NON_NATIVE_HASHER=true` env for nx tasks

JeremyJonas committed:
chore(jest): `--runInBand` when inside nx runner

JeremyJonas committed:
feat(nx-monorepo): add `linkLocalWorkspaceBins` option

This enables using workspace/sub-package bin scripct the same
way as consumers.

JeremyJonas committed:
feat(workspace): link workspace bins

JeremyJonas committed:
fix(nx-monorepo): symlink transitive bundled deps for pnpm

Add pre-package step to traverse the bundled deps transitive graph
and add symlink into packages node_modules to get them included
in pack.

JeremyJonas committed:
refactor: use linked nx-dir-hasher in diagram

agdimech commented on packages/nx-monorepo/scripts/pnpm/link-bundled-transitive-deps.ts:
```diff
@@ -0,0 +1,92 @@
+
+  const transitiveDeps: Record<string, { version: string, path: string }> = {};
+
+  function visit(_deps?: ResolvedDependencies) {
```
Why is this optional? If it needs to remain optional then we should change line 36 to: if `(!deps || _deps == null)`

JeremyJonas replied:
The conditional to check if undefined, null, or empty is inside the function, rather than have to check before each call.

JeremyJonas committed:
fix(nx-monorepo): use pnpm list instead of lock for transitive

JeremyJonas committed:
fix(nx-monorepo): remove experimental corepack suport

Related to https://github.com/aws/aws-prototyping-sdk/pull/337

agdimech commented on .husky/pre-commit:
```diff
@@ -5,9 +5,9 @@ set -eo pipefail
 
 PRE_SHA="$(git diff HEAD --ignore-space-at-eol | shasum -a 256 --tag)"
```
We should also update commit-msg to use pnpm

JeremyJonas replied:
Ack

JeremyJonas replied:
Done

agdimech commented on private/projects/pdk-monorepo-project.ts:
```diff
@@ -175,9 +172,15 @@ export class PDKMonorepoProject extends NxMonorepoProject {
-    npmrc.addConfig("yes", "true");
+    npmrc.addOverride("yes", true);
+    // https://pnpm.io/npmrc#public-hoist-pattern
+    npmrc.addOverride("public-hoist-pattern", [
```
Why do we need to hoist these?

JeremyJonas replied:
eslint and prettier are the defaults for pnpm `public-hoist-pattern`, so need append `@pnpm/*` also needed to add these back. The `@pnpm/*` was needed when I used the pnpm lock instead of pnpm list to get deps and path - but not needed now.

I will remove this chunk - not needed now.

agdimech commented on private/projects/nx-monorepo-project.ts:
```diff
@@ -17,16 +17,21 @@ export class NXMonorepoProject extends PDKProject {
     );
 
-    this.package.addBin({ "nx-dir-hasher": "./scripts/nx-dir-hasher.js" });
+    this.package.addBin({ "pdk@nx-dir-hasher": "./scripts/nx-dir-hasher.js" });
```
any reason why we are prefixing this with `pdk@`?

JeremyJonas replied:
Since these are bins the are available in PATH, and want to have clear scope for PDK based bin commands so not confused with other `nx-*` commands, and similarly with `pnpm-*` commands. The `@` is supported in filename cross-env so provides nice namespace separator.

agdimech commented on packages/nx-monorepo/scripts/pnpm/link-bundled-transitive-deps.ts:
```diff
@@ -0,0 +1,105 @@
+import * as fs from "fs-extra";
+import * as SemVer from "semver";
+
+interface Dependency {
```
Can you just use: https://github.com/pnpm/pnpm/blob/main/reviewing/list/src/getPkgInfo.ts#L13 ?

JeremyJonas replied:
Nice! Actually just found https://github.com/pnpm/pnpm/blob/main/reviewing/dependencies-hierarchy/src/buildDependenciesHierarchy.ts#L31, which is exactly what we need - no need to call `pnpm list` - we can just use it all programmatically for exactly what we need 😄

JeremyJonas replied:
Done

JeremyJonas committed:
refactor(nx-monorepo): use `@pnpm/reviewing.dependencies-hierarchy`

Use `@pnpm/reviewing.dependencies-hierarchy` instead of `pnpm list` to
get transitive bundled dependencies.

agdimech commented on packages/nx-monorepo/scripts/pnpm/link-bundled-transitive-deps.ts:
```diff
@@ -0,0 +1,90 @@
+  const pkgDir = path.join(workspaceDir, pkgFolder);
+  const pkgJson = require(path.join(pkgDir, "package.json"));
+  const bundledDeps: string[] = pkgJson.bundledDependencies || [];
+  if (bundledDeps.length < 1) {
```
NIT `bundledDeps.length == 0` ?

agdimech commented on packages/nx-monorepo/scripts/pnpm/link-bundled-transitive-deps.ts:
```diff
@@ -0,0 +1,90 @@
+  const transitiveDeps: Record<string, PackageNode> = {};
+
+  function visit(_deps?: PackageNode[]) {
+    if (_deps == null || !_deps.length) {
```
deps is optional so we also need to check for `!_deps` before checking `!_deps.length`

JeremyJonas replied:
That is what `_deps == null` is doing

agdimech replied:
_deps == null is not the same as _deps === undefined right?

JeremyJonas replied:
`x == null` is the same as `x === null || x === undefined`. The reason linters allow loose equality for `== null` use case is specifically to allow for this.

JeremyJonas replied:
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/null#difference_between_null_and_undefined

agdimech commented on packages/nx-monorepo/src/nx-monorepo.ts:
```diff
@@ -534,6 +564,64 @@ export class NxMonorepoProject extends TypeScriptProject {
+
+    if (this.package.packageManager === NodePackageManager.PNPM) {
+      // PNPM hoisting hides transitive bundled dependencies which results in
+      // transitive dependencies being packed correctly.
```
packed incorrectly*
